src/app/publications/page.tsx

```diff
@@ -1,32 +1,82 @@
 'use client'
 
+import { useState, useEffect } from 'react'
+import styled from '@emotion/styled'
+
+import { PUBLICATIONS, PublicationTypes, ResearchTopics } from '@/data/publications'
+import type { PublicationType, ResearchTopic } from '@/data/publications'
 import PublicationCard from '@/components/PublicationCard'
-import { Section, SectionContent, SectionTitle, Sections } from '@/components/Section'
-import { PUBLICATIONS } from '@/data/publications'
+import { Sections, Section, SectionTitle, SectionContent } from '@/components/Section'
+import Filter from '@/components/Filter'
 import { uniq } from 'lodash'
 
+const Filters = styled.div`
+  display: flex;
+  gap: 12px;
+`
+
 export default function Page() {
+  const [researchTopic, setResearchTopic] = useState<ResearchTopic | 'All'>('All')
+  const [publicationType, setPublicationType] = useState<PublicationType | 'All'>('All')
+  const [publicationList, setPublicationList] = useState(PUBLICATIONS)
+
+  const handleResearchTopicChange = (topic: string) => {
+    setResearchTopic(topic as ResearchTopic)
+  }
+
+  const handlePublicationTypeChange = (type: string) => {
+    setPublicationType(type as PublicationType)
+  }
+
+  useEffect(() => {
```

Comment:
This `useEffect` is for filtering publications based on the `researchTopic` and `publicationType` selected by user from dropdown menus, updating the `publicationList`. Are there any althernative approaches you would recommend instead of this way?

Reply:
Oh I thought it was a static value. But still, we can try pre-compute the list for each type.

```tsx
const publicationLists = ResearchTopics.map(topic => ({ [topic]: PUBLICATIONS.filter(p => p === topic)}))
```

If we do this, the constant should be defined outside the component or we should use `useMemo`

Reply:
I tried to use `useMemo` but I'm not sure how it works better than `useEffect` in this case. Does the following code match your intent?

```typescript
  const publicationList = useMemo(() => {
    return PUBLICATIONS.filter(
      pub =>
        (researchTopic === 'All' || pub.topics.includes(researchTopic)) &&
        (publicationType === 'All' || pub.type === publicationType)
    )
  }, [researchTopic, publicationType])
```

Reply:
`useMemo` is useless if it depends on user inputs. I suggested useMemo or constant outside of the component to prevent recomputing the list for each rendering and user input event.

Reply:
I see! So, if I understood correctly, you're suggesting to pre-compute all possible combinations at the initial rendering moment and reuse them.



```typescript
const publicationListSet = [];

ResearchTopics.forEach(
  topic => {
    PublicationTypes.forEach(
      type => {
        publicationListSet.push(
          // filtered list
          )
        );
      }
    );
  }
);

export default function Page() {
  ...
}
```

However, I thought there wouldn't be much difference in cost between pre-computing everything and computing once whenever a user applies a filter. Therefore, I'm considering sticking with `useEffect` or `useMemo`. What do you think?

Reply:
And here is how I've tried to implement the pre-computing solution.

```typescript
const publicationListSet: {
  [key: ResearchTopic | 'All']: {
    [key: PublicationType | 'All']: Publication[]
  }
} = {}

publicationListSet['All'] = {}
publicationListSet['All']['All'] = PUBLICATIONS

ResearchTopics.forEach(topic => {
  publicationListSet[topic] = {}
  publicationListSet[topic]['All'] = PUBLICATIONS.filter(pub => pub.topics.includes(topic))
  PublicationTypes.forEach(type => {
    publicationListSet['All'][type] = PUBLICATIONS.filter(pub => pub.type === type)
    publicationListSet[topic][type] = PUBLICATIONS.filter(pub => pub.topics.includes(topic) && pub.type === type)
  })
})

export default function Page() {
  ...
}
```

Reply:
Thanks a lot for your feedback! It was valuable :D